### PR TITLE
Optional decl cast (none part)

### PR DESCRIPTION
**Pull request by danieldaeschle**
```
type Foo = int | string

fn main() {
	x := ?Foo(none)
}
```

This compiles for now. But Optional doesn't have methods or attributes that are accessable..

Don't know how we can make
```
x.has_value
```
working

### CODE REVIEW

File: vlib/v/gen/c/cgen.v [hunk 1]
```diff
@@ -4403,6 +4353,66 @@ fn (mut g Gen) ident(node ast.Ident) {
 	g.write(g.get_ternary_name(name))
 }
 
+fn (mut g Gen) cast_expr(node ast.CastExpr) {
```
Comment (medvednikov) [inline vlib/v/gen/c/cgen.v:4356]: Just one nitpick: moving such logic blocks to functions is great and welcome, but it should be done in separate PRs, so that it's easy to see the diff of the actual change.

Comment (danieldaeschle) [reply vlib/v/gen/c/cgen.v:4356]: k, haven't changed anything in this case

**Comment by dumblob** [comment]
Curious what is this good for. Could someone elaborate? This feels like "reified" optionals (i.e. an optional assigned to a variable) which is something we never wanted in *V*.

**Comment by medvednikov** [comment]
@dumblob yes, it's exactly for that.

For example, with optional sum types we can now do:

```
x := ?Expr(none)
if ... {
  x = IfExpr{}
}
else {
 x = ForExpr{}
}
```

**Comment by dumblob** [comment]
Admittedly this is a serious huge change which I'd rather have reconsidered. It seems to violate 2 things:

1. The notion of reified `none` goes against the claim that *V* has no "null" (correct me if I'm wrong, but up until now it wasn't possible in *V* to assign `none` to a variable, but only `return` it).
2. Having a variable being an optional has the same cognitive burden as well as code bloat as 3-state logic which has been discussed on several places alredy and unanimously decided to be the **wrong way** to go.

I'm convinced that the example above is a niche use case (not to mention we could get away easily with the "ternary" operator with "smartcastingdoneright").

I feel `?Foo(none)` is a nasty hack to hide some *V* shortcomings instead of properly tackling them. This change makes *V* suddenly much less appealing for practical usage because it trades some seemingly niche use cases for **lots** of safety and cognitive load while programming making *V* suddenly not any more an easy to work with transparent language.

Could we elaborate more about the breadth of use cases which currently *V* doesn't support well enough and reified optionals would solve **without** side effects?

**Comment by medvednikov** [comment]
V doesn't have `null`. `none` != `null`

`none`, like all optionals must be checked. Empty value is a necessary concept.

Optionals like ?string, ?int etc are an integral part of V and vlib.  They can have empty values that must be checked. Can you elaborate how this increases the cognitive load?

**Comment by dumblob** [comment]
> V doesn't have `null`. `none` != `null`
> 
> `none`, like all optionals must be checked. Empty value is a necessary concept.

But not for variables. Empty value must be ephemeral (i.e. only "getting it" must be allowed). Otherwise see below.

> Optionals like ?string, ?int etc are an integral part of V and vlib. They can have empty values that must be checked. Can you elaborate how this increases the cognitive load?

Having a reified optional means by definition that a value can be `null` (regarding cognitive load it doesn't matter whether it's compile time `none` or runtime `null`). So if I write a function which accepts `string` like `fn f(x string){}` and the variable I'm passing to `f()` can be `none`, then I have to check it before passing it to the function which means cognitive load. Of course, I can now declare the function like `fn f(x ?string){}` but that's not better at all because I have to check for `none` in the function definition. In other words, `none` is suddenly **everywhere** - you never know where it's being used and why etc. So the mental model is exactly the same as with `null` which is bad as hell.

Whereas without reified optionals, it's **always** guaranteed it's only possible to get optional as a return value. There is absolutely no other way to get it. This significantly reduces the number of cases when you have to deal with optionals. And the mental model itself is an exact opposite to "`none` is everywhere" because "`none` can be only returned on **the way up**" which in vast majority of cases is even easy to guess because according what to the function/method does you **expect** some side effect (despite knowing what it might be) which you can easily **solve** by adding `?` or `or{println(err)}`. This is by far not true for variables because they "hold" value, but there is no way to guess whether the value is actually the value or maybe a `none`.

Therefore with reified optionals you immediately end up with three-state logic thinking, fear of working with "bombs" (`none` values), destroyed abstraction (sometimes you have to "strip optional" from a variable before passing it down a function and sometimes you have to pass an optional - and this even if you don't have any optional, so you have to convert the variable to an optional before passing), higher cognitive load, etc.

So back to my question - what are the actual **important** use cases **without** side effects for reified optionals?

Why not to rewrite the example above as something like:

```v
x := if ... {
  x = IfExpr{} in Expr // specifying that IfExpr from sumtype Expr is meant
}
else {
 x = ForExpr{} // specifying that ForExpr from the same sumtype as IfExpr is meant
}
```
or
```v
x := IfExpr{}
if ! ... {
 x = ForExpr{} in Expr // specifying that ForExpr from sumtype Expr is meant
}
```

Yes, something like that should work. And if it currently doesn't, then smartcasting or however we call it should be improved instead of adding nasty hacks which immediately open huge can of worms without any way to close it.

----

Btw. ephemeral optionals was one of the major selling point of *V* for teams working on reliable high performance SW (incl. games). Reified optionals destroy this advantage and make *V* suddenly much less appealing compared to other competing modern languages.

And also the syntax `x := ?Expr(none)` is horrible - it's not just that it resembles `null`, it's actually explicitly saying it: **assign none to x and make it compatible with Expr** (instead of *declare x as Expr and make it unusable until assigned some value*).

**Comment by medvednikov** [comment]
> Whereas without reified optionals, it's always guaranteed it's only possible to get optional as a return value. 

Ok, I understand your point, thanks. I'll think about it and do research.

**Comment by elimisteve** [comment]
> Having a variable being an optional has the same cognitive burden as well as code bloat as 3-state logic 

@dumblob `T | none` has the same cognitive burden as `T | none | error`???

**Comment by dumblob** [comment]
> > Having a variable being an optional has the same cognitive burden as well as code bloat as 3-state logic
> 
> @dumblob `T | none` has the same cognitive burden as `T | none | error`???

No. I hope this is not how my rant sounds :wink:.

I meant the simplest demonstration of 3-state logic - i.e. `?bool` versus `bool`. This PR has nothing to do with `T | none | error` as optionals are (and always were IIRC) distinct from sumtypes in *V*. *V* is thus incapable of coercing `T | none | error` to `?(T | none)`.

But making `?string` a syntactic sugar for `Error<> | string` (with `Error<>` meaning **something satisfying the Error interface** unlike `Error` which would strip off the specific error information /I deliberately didn't say "error subtype" because [I believe the rather strict subtyping can still be loosened a bit](https://github.com/vlang/v/issues/9598#issuecomment-824699700 )/) sounds acceptable to me.

**Comment by elimisteve** [comment]
@dumblob Cool :slightly_smiling_face: :+1: . Sometimes by "optionals" people mean "a value that could be either T or none" or "a value that could be either T, none, or error", so I misread you; my fault.
